test/Microsoft.IdentityModel.Tokens.Tests/EventBasedLRUCacheTests.cs

```diff
@@ -0,0 +1,242 @@
+            {
+                context.AddDiff("The key value pair {1, 'one'} should be in the cache, but the TryGetValue() method returned false.");
+                if (!value.Equals("one"))
+                    context.AddDiff("The corresponding value for key '1' should be 'one' but was '" + value + "'.");
+            }
+
+            if (cache.TryGetValue(2, out _))
+                context.AddDiff("A key value pair with a key of '2' was never added to the cache, but the TryGetValue() method returned true.");
+
+            try
+            {
+                cache.TryGetValue(null, out _);
+                context.AddDiff("The first parameter passed into the TryGetValue() method was null, but no exception was thrown.");
+            }
+            catch (Exception ex)
+            {
+                if (!ex.GetType().Equals(typeof(ArgumentNullException)))
+                    context.AddDiff("The exception type thrown by TryGetValue() was not of type ArgumentNullException.");
+            }
+
+            TestUtilities.AssertFailIfErrors(context);
+        }
+
+        [Fact]
+        public void TryRemove()
+        {
+            TestUtilities.WriteHeader($"{this}.RemoveValue");
+            var context = new CompareContext($"{this}.RemoveValue");
+            var cache = new EventBasedLRUCache<int?, string>(1);
+
+            cache.SetValue(1, "one");
+
+            if (!cache.TryRemove(1, out _))
+                context.AddDiff("The key value pair {1, 'one'} should have been removed from the cache, but the TryRemove() method returned false.");
+
+            if (cache.TryRemove(2, out _))
+                context.AddDiff("The key value pair {2, 'two'} was never added to the cache, but the TryRemove() method returned true.");
+
+            try
+            {
+                cache.TryRemove(null, out _);
+                context.AddDiff("The first parameter passed into the TryRemove() method was null, but no exception was thrown.");
+            }
+            catch (Exception ex)
+            {
+                if (!ex.GetType().Equals(typeof(ArgumentNullException)))
+                    context.AddDiff("The exception type thrown by TryRemove() was not of type ArgumentNullException.");
+            }
+
+            TestUtilities.AssertFailIfErrors(context);
+        }
+
+        [Fact]
+        public void MaintainLRUOrder()
+        {
+            TestUtilities.WriteHeader($"{this}.MaintainLRUOrder");
+            var context = new CompareContext($"{this}.MaintainLRUOrder");
+            var cache = new EventBasedLRUCache<int, string>(10);
+
+            for (int i = 0; i <= 100; i++)
+            {
+                cache.SetValue(i, Guid.NewGuid().ToString());
+
+                // print the value of the list every 100 items
+                if (i % 10 == 0 && i != 0)
+                {
+                    // wait for the cache events to process
+                    Thread.Sleep(1000);
```

Comment:
>Thread.Sleep(1000); [](start = 20, length = 19)

perhaps the cache should have an event for events processed? avoiding the needed for a thread.sleep? Or alternatively a WaitForProcessing on the LRUCache?

Reply:
I do think this is a good idea. I think the simplest way of doing this would be to add a bool flag (e.g. 'IsBusy') that tells the user whether the _eventQueue is empty or not, but this would require the user to figure out a way to wait until this condition is false. If we're only adding this for the purpose of testing I think this solution would be fine, but if we expect to expose it publicly having an event be fired or creating a WaitForProcessing() method seems more ideal.

I'm not too familiar with how I'd go about implementing something like this; are there any best practices that you're aware of?

Reply:
Thought: Instead of a timer. Move the "wait" to a method on the cache. The method can force push queue processing (internal method)